File: python/sorting_algorithms/algos/pigeonhole_sort.py
from typing import Sequence, Optional, List
from ..base import _setup_tracer, _finalize, SortTracer, Number


def pigeonhole_sort(data: Sequence[int], *, trace: bool = False, tracer: Optional[SortTracer] = None):
    arr = list(data)
    tracer, ret_trace = _setup_tracer("pigeonhole_sort", trace, tracer)
    if not arr:
        return _finalize(arr, tracer, ret_trace)
    min_val, max_val = min(arr), max(arr)
    size = max_val - min_val + 1
    holes = [0] * size
    for x in arr:
        holes[x - min_val] += 1
        if tracer:
            tracer.record(arr, (), f"count {x}")
    idx = 0
    for i, count in enumerate(holes):
        while count > 0:
            arr[idx] = i + min_val
            idx += 1
            count -= 1
            if tracer:
                tracer.record(arr, (idx - 1,), "fill")
    return _finalize(arr, tracer, ret_trace)

File: c/src/pigeonhole_sort.c
#include "../sorting_algorithms.h"
#include <stdlib.h>
#include <string.h>

void pigeonhole_sort(int *arr, size_t n) {
    if (n == 0) return;
    int min = arr[0], max = arr[0];
    for (size_t i = 1; i < n; ++i) {
        if (arr[i] < min) min = arr[i];
        if (arr[i] > max) max = arr[i];
    }
    size_t range = (size_t)(max - min + 1);
    int *holes = (int *)calloc(range, sizeof(int));
    if (!holes) return;
    for (size_t i = 0; i < n; ++i) holes[arr[i] - min]++;
    size_t idx = 0;
    for (size_t i = 0; i < range; ++i) {
        while (holes[i]-- > 0) arr[idx++] = (int)(i + min);
    }
    free(holes);
}
